c7n/resources/directory.py

```diff
@@ -122,3 +151,58 @@ class resource_type(TypeInfo):
         universal_taggable = object()
 
     augment = universal_augment
+
+@CloudDirectory.action_registry.register('delete')
+class CloudDirectoryDelete(BaseAction):
+    """Delete a cloud directory.
+
+    .. code-block:: yaml
+
+       policies:
+         - name: delete-cloud-directory
+           resource: aws.cloud-directory
+           filters:
+             - Name: test-cloud
+           actions:
+             - type: delete
+    """
+    schema = type_schema('delete')
+    permissions = ('clouddirectory:DeleteDirectory',
+                   'clouddirectory:DisableDirectory',)
+
+    def process(self, resources):
+        client = local_session(
+            self.manager.session_factory).client('clouddirectory')
+        for r in resources:
```

Comment:
Since it looks like we can only delete directories that are disabled, we may want to explicitly skip directories that aren't in that state:

```suggestion
        self.filter_resources(resources, 'State', ('DISABLED',))
        for r in resources:
```

We do something similar [with ec2 terminations](https://github.com/cloud-custodian/cloud-custodian/blob/13c3768760740e292d78b03e7562e4c145fb9175/c7n/resources/ec2.py#L1663).

Another option might be to add a `force` option that would disable directories as needed, similar to how we can [opt into some cleanup steps](https://github.com/cloud-custodian/cloud-custodian/blob/13c3768760740e292d78b03e7562e4c145fb9175/c7n/resources/iam.py#L1384-L1386) when deleting a role.

While testing these policies, did you run into any long delays (more than a few seconds) between disabling a directory and being able to delete it?

Reply:
Hi @ajkerrigan, thanks for the review! On [line 118](https://github.com/cloud-custodian/cloud-custodian/blob/e037b6b8efb59fe95e2f4ca13311208bc5e1829f/c7n/resources/directory.py#L118), it seems to filter on `ENABLED` resources, which is why I didn't do a `force` option.

As for long delays between api calls, I did not but will switch to using two loops to prevent this if it happens. 
Thanks!

Reply:
> Hi @ajkerrigan, thanks for the review! On [line 118](https://github.com/cloud-custodian/cloud-custodian/blob/e037b6b8efb59fe95e2f4ca13311208bc5e1829f/c7n/resources/directory.py#L118), it seems to filter on `ENABLED` resources, which is why I didn't do a `force` option.

Oh good catch I missed that, interesting! Hmm that suggests that the `delete` action wouldn't ever work unless we had `ENABLED` resources in the local Custodian cache that were disabled. Because otherwise it wouldn't fetch the disabled resources.

In that case, assuming we still want to fetch only enabled directories by default... it might make sense to opt into showing disabled ones with a query. Similar to how `aws.ebs` fetches only your own snapshots by default, but lets you customize the `query` option to consider other owners: https://github.com/cloud-custodian/cloud-custodian/blob/5fd45ee11710cd3deea791a2ac9d09d9f7189005/c7n/resources/ebs.py#L71-L72